rescompy/features/features.py

```diff
@@ -200,9 +203,448 @@ def __call__(self, r, u):
             s = np.concatenate((s, r**poly_ind), axis=1)
         return s
 
-    def feature_size(self, esn_size:int,input_dim:int): 
+    def feature_size(self, esn_size:int, input_dim:int): 
         return 1+input_dim+esn_size*self.degree
 
+    def jacobian(r: np.ndarray, u: np.ndarray, dr_du: np.ndarray):
+        raise NotImplementedError()
+
+@dataclass
+class FinalStateOnly(ESNFeature):
+    """The Final-state-only Feature function.
+    
+    Simply returns the final reservoir state of a driving period.
+    
+    Args:
+        r (np.ndarray): The reservoir states.
+        u (np.ndarray): The input signal.
+        
+    Returns:
+        s (np.ndarray): The feature vector.
+    """
+
+    @staticmethod
+    def __call__(r : np.ndarray, u : np.ndarray):
+        r = r.reshape((-1, r.shape[-1]))
+        s = np.copy(r[-1].reshape(1, -1)) 
+        return s
+	
+    def __init__(self):
+        self.compiled = numba.njit(lambda r, u: np.copy(r.reshape((-1, r.shape[-1]))[-1].reshape(1, -1)))
 
+    @staticmethod	
+    def feature_size(esn_size: int, signal_length: int):
+        return esn_size
+
+    @staticmethod
     def jacobian(r: np.ndarray, u: np.ndarray, dr_du: np.ndarray):
         raise NotImplementedError()
+
```

Comment:
I left the MRS and lookback-related features as functions rather than classes so that they can be jit-compiled. Since they can't be written as lambda functions, right now I don't see a way to give a compiled option within a class definition. The jit-compilation is useful for the MRS feature function since it is called thousands (or tens of thousands) of times in the training stage.

That said, ESNFeature class versions of the MRS and lookback-related features are commented out at the bottom of the script. I could make them the default and just use my own jit-compiled version when I want to generate results.

Reply:
gotcha.  you could do something like:

```
def __init__(self):
  def inner(args):
      # do something
  self.compiled = numba.njit(inner)
```

Reply:
Works like a charm! Thanks. Added it to the polynomial ESNFeature too.